src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/Filterator.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.graphanalysis;
+
+import com.google.common.base.Predicate;
+
+import javax.annotation.Nonnull;
+import java.util.Iterator;
+
+/** Iterator that allows returned objects to be filtered against a given condition
+ *  @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public interface Filterator<T> extends Iterator<T> {
+    /** Returns a filtered view of an iterable */
+    @Nonnull
+    public Filterator<T> filter(@Nonnull Predicate<T> matchCondition);
```

Comment:
Would be clear if this extended `Iterable`, but unclear what this means as a kind of `Iterator`. Is the return value expected to continue the iteration from the same spot as the original? Can the original still be used after this call, or is that an error? Is `remove` supported?